nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java

```diff
@@ -0,0 +1,290 @@
+    }
+
+    private Schema createAndLoadSchema(final Path tempDir) {
+        final SchemaLoader schemaLoader = new SchemaLoader(FileSystems.getDefault());
+
+        final List<Location> roots = new ArrayList<>();
+        roots.add(Location.get(tempDir.toString()));
+
+        // Add standard protobuf libraries
+        roots.addAll(STANDARD_PROTOBUF_LOCATIONS);
+
+        schemaLoader.initRoots(roots, Collections.emptyList());
+
+        // Load and return the compiled schema
+        return schemaLoader.loadSchema();
+    }
+
+    private void safeDeleteDirectory(final Path directory) {
+        if (Files.exists(directory)) {
+            try {
+                FileUtils.deleteDirectory(directory.toFile());
+            } catch (final IOException e) {
+                logger.warn("Failed to delete temporary directory: {}", directory, e);
+            }
+        }
+    }
+
+    /**
+     * Validates that all SchemaDefinition identifiers end with .proto extension.
+     * Performs recursive validation on all referenced schemas.
+     *
+     * @param schemaDefinition       the schema definition to validate
+     * @param isRootSchemaDefinition set to if schema definition is a root definition
+     * @throws IllegalArgumentException if any identifier does not end with .proto extension
+     */
+    public void validateSchemaDefinitionIdentifiers(final SchemaDefinition schemaDefinition, final boolean isRootSchemaDefinition) {
+        // do not validate schema identifier names for root schema definitions. They might be coming from sources like text fields,
+        // flow file attributes and other sources that do not support naming.
+        if (!isRootSchemaDefinition) {
+            validateSchemaIdentifier(schemaDefinition.getIdentifier());
+        }
+
+        // Recursively validate all referenced schemas
+        // schema references have to end with .proto extension.
+        for (final SchemaDefinition referencedSchema : schemaDefinition.getReferences().values()) {
+            validateSchemaDefinitionIdentifiers(referencedSchema, false);
+        }
+    }
+
+    /**
+     * Validates that a single SchemaIdentifier has a name ending with .proto extension.
+     *
+     * @param schemaIdentifier the schema identifier to validate
+     * @throws IllegalArgumentException if the identifier name does not end with .proto extension
+     */
+    private void validateSchemaIdentifier(final SchemaIdentifier schemaIdentifier) {
+        final Optional<String> nameOptional = schemaIdentifier.getName();
+        if (nameOptional.isPresent()) {
+            final String name = nameOptional.get();
+            if (!name.endsWith(".proto")) {
+                throw new IllegalArgumentException(
+                    String.format("Schema identifier name '%s' must end with .proto extension", name));
+            }
+        } else {
+            throw new IllegalArgumentException("Schema identifier must have a name that ends with .proto extension");
```

Comment:
A little too ambiguous for my taste
```suggestion
            throw new IllegalArgumentException("Schema identifier must have a name that ends with .proto extension, but it has none");
```